examples/WGAN-GP/model.py

```diff
@@ -0,0 +1,170 @@
+            )(x)
+            x = elegy.nn.BatchNormalization(decay_rate=0.9)(x)
+            x = jax.nn.leaky_relu(x, negative_slope=0.2)
+        x = elegy.nn.conv.Conv2DTranspose(3, (4, 4), stride=(2, 2))(x)
+        x = jax.nn.sigmoid(x)
+        return x
+
+
+# the discriminator architecture adapted from DCGAN
+# also called 'critic' in the WGAN paper
+class Discriminator(elegy.Module):
+    def call(self, x):
+        for c in [128, 256, 512, 1024]:
+            x = elegy.nn.conv.Conv2D(c, (4, 4), stride=(2, 2))(x)
+            x = jax.nn.leaky_relu(x, negative_slope=0.2)
+        x = elegy.nn.Flatten()(x)
+        x = elegy.nn.Linear(1)(x)
+        return x
+
+
+# multiplier for gradient normalization
+LAMBDA_GP = 10
+
+# gradient regularization term
+def gradient_penalty(x_real, x_fake, applied_discriminator_fn, rngkey):
+    assert len(x_real) == len(x_fake)
+    alpha = jax.random.uniform(rngkey, shape=[len(x_real), 1, 1, 1])
+    x_hat = x_real * alpha + x_fake * (1 - alpha)
+    grads = jax.grad(lambda x: applied_discriminator_fn(x)[0].mean())(x_hat)
+    norm = jnp.sqrt((grads ** 2).sum(axis=[1, 2, 3]))
+    penalty = (norm - 1) ** 2
+    return penalty.mean() * LAMBDA_GP
+
+
+class WGAN_GP(elegy.Model):
+    def __init__(self):
+        # run_eagerly=True is needed to train the generator only every 5 iterations
+        # as recommended in the WGAN paper
+        super().__init__(run_eagerly=True)
+        self.generator = Generator()
+        self.discriminator = Discriminator()
+        self.g_optimizer = optax.adam(2e-4, b1=0.5)
+        self.d_optimizer = optax.adam(2e-4, b1=0.5)
+
+        # iteration counter
+        self.i = 0
+
+    def init(self, x):
+        rng = elegy.RNGSeq(0)
+        gx, g_params, g_states = self.generator.init(rng=rng)(x)
+        dx, d_params, d_states = self.discriminator.init(rng=rng)(gx)
+
+        g_optimizer_states = self.g_optimizer.init(g_params)
+        d_optimizer_states = self.d_optimizer.init(d_params)
+
+        self.states = elegy.States(
+            g_states=g_states,
+            d_states=d_states,
+            g_params=g_params,
+            d_params=d_params,
+            g_opt_states=g_optimizer_states,
+            d_opt_states=d_optimizer_states,
+            rng=rng,
+        )
+
+    def pred_step(self, x, states):
+        z = x
+        x_fake = self.generator.apply(states.g_params, states.g_states)(z)[0]
+        return (x_fake, states)
+
+    def train_step(self, x, states):
+        # training the discriminator on every iteration
+        d_loss, d_params, d_states, d_opt_states, rng, gp = self.discriminator_step_jit(
+            x, **states
+        )
+        states = states.update(
+            d_params=d_params, d_states=d_states, d_opt_states=d_opt_states, rng=rng
+        )
+
+        self.i += 1
+        # training the generator only every 5 iterations as recommended in the original WGAN paper
+        if self.i % 5 == 0:
```

Comment:
An alternative strategy to avoid running this eagerly could be do do something like this:

```python
def do_update():
        g_loss, g_params, g_states, g_opt_states, rng = self.generator_step(
        len(x), **states
    )
    states = states.update(
        g_params=g_params, g_states=g_states, g_opt_states=g_opt_states, rng=rng
    )
    logs = {"d_loss": d_loss, "g_loss": g_loss, "gp": gp}
    return logs, states

def no_update():
    logs = {"d_loss": d_loss, "g_loss": 0.0, "gp": gp}
    return logs, states

logs, states = jax.lax.cond(states.step % 5 == 0, do_update, no_update)

states = states.update(step=states.step + 1)

return logs, states
```

Reply:
I had actually tried that before, but got very long tracing times and even a segfault. I've now updated jax, and it seems to work, maybe there was a problem with the previous version